pkg/ds/interfaces.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+// Package ds defines and implements a collection of data structures,
+// designed to hold cross-products of two Set structures in a succinct and canonical way.
+// The most important interface is Product. It is implemented by ProductLeft using injective
+// (one-to-one) mapping from sets to sets, where each key-value pair defines a complete cross
+// product of the two sets.
+package ds
+
+type Sized interface {
+	IsEmpty() bool
+	// Size returns the actual, full size of the set. For Product, it returns the number of pairs.
+	Size() int
+}
+
+type Comparable[Self any] interface {
+	Equal(Self) bool
+	Copy() Self
+}
+
+type Hashable[Self any] interface {
+	Comparable[Self]
+	Hash() int
+}
+
+type Set[Self any] interface {
+	Hashable[Self]
+	Sized
+	IsSubset(Self) bool
+	Union(Self) Self
+	Intersect(Self) Self
+	Subtract(Self) Self
+}
+
+// Product is a cartesian product of sets S1 x S2
+type Product[S1 Set[S1], S2 Set[S2]] interface {
+	Set[Product[S1, S2]]
+	Partitions() []Pair[S1, S2]
```

Comment:
```suggestion
        // Partitions returns a slice of pairs in the product set  (note that the order is arbitrary)
	Partitions() []Pair[S1, S2]
```

Reply:
This is incorrect. It returns a slice of pairs, where the union of their cross product contains exactly the same pairs as this `Product`.

Reply:
Docs are updated:
```golang
// Partitions returns a slice of pairs such that, for (p Product):
// 	  p.Equal(Union(CartesianPairLeft(s1, s2) for _, (s1, s2) := range p.Partitions())
// (note that the order is arbitrary; we do not return HashSet because Pair is not Hashable)
// Partitions returns a slice of pairs in the product set  (note that the order is arbitrary)
```